crates/hulk/src/nao/double_buffered_reader.rs

```diff
@@ -0,0 +1,411 @@
+use std::{
+    io::{self, ErrorKind, Read},
+    mem::size_of,
+    os::unix::{io::AsRawFd, prelude::RawFd},
+};
+
+use super::hula::StateStorage;
+
+pub struct DoubleBufferedReader {
+    buffers: [[u8; size_of::<StateStorage>()]; 2],
+    active_buffer_index: usize,
+    incomplete_offset: usize,
+}
+
+impl Default for DoubleBufferedReader {
+    fn default() -> Self {
+        Self {
+            buffers: [[0; size_of::<StateStorage>()]; 2],
+            active_buffer_index: Default::default(),
+            incomplete_offset: Default::default(),
+        }
+    }
+}
+
+impl DoubleBufferedReader {
+    fn swap_buffers(&mut self) {
+        self.incomplete_offset = 0;
+        self.active_buffer_index = (self.active_buffer_index + 1) % 2;
+    }
+
+    pub fn read(
```

Comment:
```suggestion
    pub fn draining_read(
```

??? What do you think?

Reply:
I called it not `drain()`. I'm not 100% happy with this name because it actually drains and returns the last complete value. `drain()` does not reflect the value returning but maybe that is okay.

Reply:
but `draining_read` does? What is your objection to that name?

Reply:
Renamed to `draining_read()`.